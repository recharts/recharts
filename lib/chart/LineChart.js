'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _generateCategoricalChart = require('./generateCategoricalChart');

var _generateCategoricalChart2 = _interopRequireDefault(_generateCategoricalChart);

var _Line = require('../cartesian/Line');

var _Line2 = _interopRequireDefault(_Line);

var _XAxis = require('../cartesian/XAxis');

var _XAxis2 = _interopRequireDefault(_XAxis);

var _YAxis = require('../cartesian/YAxis');

var _YAxis2 = _interopRequireDefault(_YAxis);

var _CartesianUtils = require('../util/CartesianUtils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = (0, _generateCategoricalChart2.default)({
  chartName: 'LineChart',
  GraphicalChild: _Line2.default,
  axisComponents: [{ axisType: 'xAxis', AxisComp: _XAxis2.default }, { axisType: 'yAxis', AxisComp: _YAxis2.default }],
  formatAxisMap: _CartesianUtils.formatAxisMap
}); /**
     * @fileOverview Line Chart
     */