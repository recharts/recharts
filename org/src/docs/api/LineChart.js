export default {
  name: 'LineChart',
  props: [
    {
      name: 'layout',
      type: "'horizontal' | 'vertical'",
      defaultVal: "'horizontal'",
      isOptional: false,
      desc: {
        'en-US': 'The layout of lines in the chart.',
        'zh-CN': '布局类型，是横向的还是纵向的。',
      },
    },
    {
      name: 'syncId',
      type: 'String',
      isOptional: true,
      desc: {
        'en-US':
          'If any two categorical charts(LineChart, AreaChart, BarChart, ComposedChart) have the same syncId, these two charts can sync the position tooltip, and the startIndex, endIndex of Brush.',
        'zh-CN':
          '如果任何两个或者多个类目型图表（LineChart, AreaChart, BarChart, ComposedChart）具有相同的 syncId，这两个图表可以联动，包括 Tooltip 以及 Brush 的联动。',
      },
      examples: [
        {
          name: 'Two synchronized LineChart and an AreaChart',
          url: '/examples/SynchronizedLineChart',
        },
      ],
    },
    {
      name: 'syncMethod',
      type: "'index' | 'value' | function",
      defaultVal: "'index'",
      isOptional: true,
      desc: {
        'en-US':
          "When syncId is provided, allows customisation of how the charts will synchronize tooltips and brushes. Using 'index' (default setting), other charts will reuse current datum's index within the data array. In cases where data does not have the same length, this might yield unexpected results. In that case use 'value' which will try to match other charts values, or a fully custom function which will receive tick, data as argument and should return an index."
      },
    },
    {
      name: 'width',
      type: 'Number',
      defaultVal: 'undefined',
      isOptional: false,
      desc: {
        'en-US': 'The width of chart container.',
        'zh-CN': '图表的宽度。',
      },
    },
    {
      name: 'height',
      type: 'Number',
      defaultVal: 'undefined',
      isOptional: false,
      desc: {
        'en-US': 'The height of chart container.',
        'zh-CN': '图表的高度。',
      },
    },
    {
      name: 'data',
      type: 'Array',
      defaultVal: 'undefined',
      isOptional: false,
      desc: {
        'en-US': 'The source data, in which each element is an object.',
        'zh-CN': '输入数据，现在支持的类型是对象数组。',
      },
      format: ["[{name: 'a', value: 12}]"],
    },
    {
      name: 'margin',
      type: 'Object',
      defaultVal: '{ top: 5, right: 5, bottom: 5, left: 5 }',
      isOptional: false,
      desc: {
        'en-US': 'The sizes of whitespace around the container.',
        'zh-CN': '图表四周的留白大小，支持传入部分值（如： { top: 5 }）',
      },
      format: ['{ top: 5, right: 5, bottom: 5, left: 5 }'],
    },
    {
      name: 'onClick',
      type: 'Function',
      isOptional: true,
      desc: {
        'en-US': 'The customized event handler of click in this chart.',
        'zh-CN': '鼠标在图表图形区域点击事件的回调函数。',
      },
    },
    {
      name: 'onMouseEnter',
      type: 'Function',
      isOptional: true,
      desc: {
        'en-US': 'The customized event handler of mouseenter in this chart.',
        'zh-CN': '鼠标在图表图形区域 mouseenter 事件的回调函数。',
      },
    },
    {
      name: 'onMouseMove',
      type: 'Function',
      isOptional: true,
      desc: {
        'en-US': 'The customized event handler of mousemove in this chart.',
        'zh-CN': '鼠标在图表图形区域 mousemove 事件的回调函数。',
      },
    },
    {
      name: 'onMouseLeave',
      type: 'Function',
      isOptional: true,
      desc: {
        'en-US': 'The customized event handler of mouseleave in this chart.',
        'zh-CN': '鼠标在图表图形区域 mouseleave 事件的回调函数。',
      },
    },
  ],
  parentComponents: ['ResponsiveContainer'],
  childrenComponents: [
    'XAxis',
    'YAxis',
    'ReferenceArea',
    'ReferenceDot',
    'ReferenceLine',
    'Brush',
    'CartesianGrid',
    'Legend',
    'Tooltip',
    'Line',
    'Customized',
    'validate svg elements...',
  ],
};
