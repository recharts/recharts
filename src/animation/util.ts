/*
 * @description: convert camel case to dash case
 * string => string
 */
export const getDashCase = (name: string) => name.replace(/([A-Z])/g, v => `-${v.toLowerCase()}`);

export const getTransitionVal = (props: ReadonlyArray<string>, duration: string | number, easing: string): string =>
  props.map(prop => `${getDashCase(prop)} ${duration}ms ${easing}`).join(',');

/**
 * Finds the intersection of keys between two objects
 * @param {object} preObj previous object
 * @param {object} nextObj next object
 * @returns an array of keys that exist in both objects
 */
export const getIntersectionKeys = (preObj: Record<string, unknown>, nextObj: Record<string, unknown>): string[] =>
  [Object.keys(preObj), Object.keys(nextObj)].reduce((a, b) => a.filter(c => b.includes(c)));

/**
 * Maps an object to another object
 * @param {function} fn function to map
 * @param {object} obj object to map
 * @returns mapped object
 */
export const mapObject = <T extends Record<string, any>, R>(
  fn: (key: keyof T, value: T[keyof T]) => R,
  obj: T,
): { [K in keyof T]: R } =>
  Object.keys(obj).reduce(
    (res, key) => ({
      ...res,
      [key]: fn(key as keyof T, obj[key as keyof T]),
    }),
    {} as { [K in keyof T]: R },
  );
