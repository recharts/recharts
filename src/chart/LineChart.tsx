/**
 * @fileOverview Line Chart
 */
import { generateCategoricalChart } from './generateCategoricalChart';
import { arrayTooltipSearcher } from '../state/optionsSlice';

export const LineChart = generateCategoricalChart({
  chartName: 'LineChart',
  tooltipPayloadSearcher: arrayTooltipSearcher,
});
