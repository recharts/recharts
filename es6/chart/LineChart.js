/**
 * @fileOverview Line Chart
 */
import { generateCategoricalChart } from './generateCategoricalChart';
import { Line } from '../cartesian/Line';
import { XAxis } from '../cartesian/XAxis';
import { YAxis } from '../cartesian/YAxis';
import { formatAxisMap } from '../util/CartesianUtils';
export var LineChart = generateCategoricalChart({
  chartName: 'LineChart',
  GraphicalChild: Line,
  axisComponents: [{
    axisType: 'xAxis',
    AxisComp: XAxis
  }, {
    axisType: 'yAxis',
    AxisComp: YAxis
  }],
  formatAxisMap: formatAxisMap
});