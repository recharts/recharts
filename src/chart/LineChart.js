/**
 * @fileOverview Line Chart
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Smooth from 'react-smooth';
import _ from 'lodash';
import classNames from 'classnames';
import Layer from '../container/Layer';
import Tooltip from '../component/Tooltip';
import Curve from '../shape/Curve';
import Dot from '../shape/Dot';
import generateCategoricalChart from './generateCategoricalChart';
import Line from '../cartesian/Line';
import { getPresentationAttributes, findChildByType } from '../util/ReactUtils';
import AnimationDecorator from '../util/AnimationDecorator';
import composedDataDecorator from '../util/ComposedDataDecorator';
import { getValueByDataKey } from '../util/DataUtils';

const getCategoryAxisCoordinate = ({ axis, ticks, bandSize, entry, index }) => {
  if (axis.type === 'category') {
    return ticks[index] ? ticks[index].coordinate + bandSize / 2 : null;
  }

  const value = getValueByDataKey(entry, axis.dataKey);

  return !_.isNil(value) ? axis.scale(value) : null;
};
/**
 * Compose the data of each group
 * @param {Object} props The props from the component
 * @param  {Object} xAxis   The configuration of x-axis
 * @param  {Object} yAxis   The configuration of y-axis
 * @param  {String} dataKey The unique key of a group
 * @return {Array}  Composed data
 */
const getComposedData = ({ props, xAxis, yAxis, xTicks, yTicks, dataKey, bandSize }) => {
  const { layout, dataStartIndex, dataEndIndex } = props;
  const data = props.data.slice(dataStartIndex, dataEndIndex + 1);

  return data.map((entry, index) => {
    const value = getValueByDataKey(entry, dataKey);

    if (layout === 'horizontal') {
      return {
        x: getCategoryAxisCoordinate({ axis: xAxis, ticks: xTicks, bandSize, entry, index }),
        y: _.isNil(value) ? null : yAxis.scale(value),
        value,
        payload: entry,
      };
    }

    return {
      x: _.isNil(value) ? null : xAxis.scale(value),
      y: getCategoryAxisCoordinate({ axis: yAxis, ticks: yTicks, bandSize, entry, index }),
      value,
      payload: entry,
    };
  });
};

@AnimationDecorator
@composedDataDecorator({ getComposedData })
export class LineChart extends Component {

  static displayName = 'LineChart';

  static propTypes = {
    allComposedData: PropTypes.array,
    axisTicks: PropTypes.array,
    layout: PropTypes.oneOf(['horizontal', 'vertical']),
    dataStartIndex: PropTypes.number,
    dataEndIndex: PropTypes.number,
    data: PropTypes.array,
    isTooltipActive: PropTypes.bool,
    activeTooltipIndex: PropTypes.number,
    xAxisMap: PropTypes.object,
    yAxisMap: PropTypes.object,
    offset: PropTypes.object,
    graphicalItems: PropTypes.array,
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node,
    ]),
    // used internally
    isComposed: PropTypes.bool,
    animationId: PropTypes.number,
  };

  renderCursor({ offset }) {
    const { children, isTooltipActive, layout, activeTooltipIndex, axisTicks } = this.props;
    const tooltipItem = findChildByType(children, Tooltip);

    if (!tooltipItem || !tooltipItem.props.cursor || !isTooltipActive ||
      activeTooltipIndex < 0) { return null; }

    const ticks = axisTicks;

    if (!ticks || !ticks[activeTooltipIndex]) { return null; }

    const start = ticks[activeTooltipIndex].coordinate;
    const x1 = layout === 'horizontal' ? start : offset.left;
    const y1 = layout === 'horizontal' ? offset.top : start;
    const x2 = layout === 'horizontal' ? start : offset.left + offset.width;
    const y2 = layout === 'horizontal' ? offset.top + offset.height : start;
    const cursorProps = {
      stroke: '#ccc',
      ...getPresentationAttributes(tooltipItem.props.cursor),
      points: [{ x: x1, y: y1 }, { x: x2, y: y2 }],
    };

    return React.isValidElement(tooltipItem.props.cursor) ?
      React.cloneElement(tooltipItem.props.cursor, cursorProps) :
      <Curve {...cursorProps} type="linear" className="recharts-tooltip-cursor" />;
  }

  renderActiveDot(option, props, childIndex) {
    let dot;

    if (React.isValidElement(option)) {
      dot = React.cloneElement(option, props);
    } else if (_.isFunction(option)) {
      dot = option(props);
    } else {
      const className = classNames('recharts-line-active-dot', option.className);
      dot = <Dot {...props} className={className} />;
    }

    return (
      <Smooth
        from="scale(0)"
        to="scale(1)"
        duration={400}
        key={`dot-${childIndex}`}
        attributeName="transform"
      >
        <Layer style={{ transformOrigin: 'center center' }}>{dot}</Layer>
      </Smooth>
    );
  }
  /**
   * Draw the main part of line chart
   * @param  {Array} items     All the instance of Line
   * @param  {Object} xAxisMap The configuration of all x-axes
   * @param  {Object} yAxisMap The configuration of all y-axes
   * @param  {Object} offset   The offset of main part in the svg element
   * @return {ReactComponent}  All the instances of Line
   */
  renderItems(items, xAxisMap, yAxisMap, offset) {
    const { children, layout, isTooltipActive, activeTooltipIndex,
      animationId, allComposedData } = this.props;
    const tooltipItem = findChildByType(children, Tooltip);
    const hasDot = tooltipItem && isTooltipActive;
    const dotItems = [];

    const lineItems = items.map((child, i) => {
      const { dataKey, stroke, activeDot } = child.props;
      const points = allComposedData[i];
      const activePoint = points[activeTooltipIndex];

      if (hasDot && activeDot && activePoint) {
        const dotProps = {
          index: activeTooltipIndex,
          dataKey,
          cx: activePoint.x, cy: activePoint.y, r: 4,
          fill: stroke, strokeWidth: 2, stroke: '#fff',
          payload: activePoint.payload,
          value: activePoint.value,
          ...getPresentationAttributes(activeDot),
        };
        dotItems.push(this.renderActiveDot(activeDot, dotProps, i));
      }

      return React.cloneElement(child, {
        key: child.key || `line-${i}`,
        ...offset,
        layout,
        points,
        animationId,
        xAxis: xAxisMap[child.props.xAxisId],
        yAxis: yAxisMap[child.props.yAxisId],
      });
    }, this);

    return (
      <Layer className="recharts-line-chart-wrapper">
        <Layer className="recharts-line-chart">{lineItems}</Layer>
        <Layer className="recharts-line-chart-dot">{dotItems}</Layer>
      </Layer>
    );
  }

  render() {
    const { isComposed, xAxisMap, yAxisMap, offset, graphicalItems } = this.props;

    return (
      <Layer className="recharts-line-graphical">
        {!isComposed && this.renderCursor({ offset })}
        {this.renderItems(graphicalItems, xAxisMap, yAxisMap, offset)}
      </Layer>
    );
  }
}

export default generateCategoricalChart(LineChart, Line);
