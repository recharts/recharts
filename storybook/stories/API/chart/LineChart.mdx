import { Canvas, Meta, ArgTypes } from '@storybook/blocks';
import * as LineChartStories from './LineChart.stories';

# LineChart

## Parent Component

The LineChart can be used within: `<ResponsiveContainer />`.

## Child component

The LineChart can be used with the following child components: `<XAxis />`, `<YAxis />`, `<ReferenceArea />`, `<ReferenceDot />`, `<ReferenceLine />`, `<Brush />`, `<CartesianGrid />`, `<Legend />`, `<Tooltip />`, `<Line />`, `<Customized />` or valid svg elements.

## Properties

Properties in the groups Other and Internal are not recommended to be used.

<ArgTypes of={LineChartStories} sort={'requiredFirst'} />
