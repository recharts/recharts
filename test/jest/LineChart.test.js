/**
 * @jest-environment jsdom
 */
import React from 'react';
import { LineChartWrapper, handleMouseOver, handleSyncMethod } from './data/LineChart.data';
import { setup } from './utils/setup.util';


const {user, container} = setup(<LineChartWrapper />);

test('Can create a test snapshot', async () => {
  expect(container).toMatchSnapshot();
});

test('A cursor does not exist', async () => {
  expect(container.querySelector('recharts-tooltip-cursor')).toBeNull;
});

test('Tooltip wrapper exists', () => {
  expect(container.querySelector('recharts-tooltip-wrapper')).not.toBeNull;
});

test('Tooltips are not visible by default', () => {
  expect(container.querySelector('recharts-tooltip-wrapper')).not.toBeVisible;
})

test('Both tooltips are visible on hover', () => {
  const element = container.querySelector('recharts-tooltip-wrapper');

  user.hover(element)
      .then(() => {
        const tooltipElements = container.querySelectorAll('recharts-tooltip-wrapper');

        tooltipElements.forEach((tooltipElement) => {
          expect(tooltipElement).toBeVisible
        });

        expect(handleMouseOver.mock.calls.length).toBe(2);
        expect(handleSyncMethod.mock.calls.length).toBe(2);
      })
});

test('Correct values are displayed on tooltips', () => {
  const element = container.querySelector('recharts-tooltip-wrapper');

  user.hover(element)
      .then(() => {
        const tooltipElements = container.querySelectorAll('recharts-tooltip-item-value');

        expect(tooltipElements[0].props().value).toBe(400);
        expect(tooltipElements[1].props().value).toBe(550);
      })
});

