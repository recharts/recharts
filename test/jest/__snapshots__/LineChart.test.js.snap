// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can create a test snapshot 1`] = `
<div>
  <div>
    <div
      class="recharts-wrapper"
      style="position: relative; cursor: default; width: 400px; height: 400px;"
    >
      <svg
        class="recharts-surface"
        height="400"
        version="1.1"
        viewBox="0 0 400 400"
        width="400"
      >
        <defs>
          <clippath
            id="recharts1-clip"
          >
            <rect
              height="330"
              width="360"
              x="20"
              y="20"
            />
          </clippath>
        </defs>
        <g
          class="recharts-layer recharts-line"
        >
          <path
            class="recharts-curve recharts-line-curve"
            d="M20,20C44,61.25,68,102.5,92,102.5C116,102.5,140,102.5,164,102.5C188,102.5,212,185,236,185C260,185,284,120.65,308,120.65C332,120.65,356,157.3625,380,194.07500000000002"
            fill="none"
            height="330"
            points="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
            stroke="#ff7300"
            stroke-dasharray="0px 0px"
            stroke-width="1"
            type="monotone"
            width="360"
          />
        </g>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
        >
          <line
            class="recharts-cartesian-axis-line"
            fill="none"
            height="30"
            orientation="bottom"
            stroke="#666"
            type="category"
            width="360"
            x="20"
            x1="20"
            x2="380"
            y="350"
            y1="350"
            y2="350"
          />
          <g
            class="recharts-cartesian-axis-ticks"
          >
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                type="category"
                width="360"
                x="20"
                x1="20"
                x2="20"
                y="350"
                y1="356"
                y2="350"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                type="category"
                width="360"
                x="20"
                y="358"
              >
                <tspan
                  dy="0.71em"
                  x="20"
                >
                  Page A
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                type="category"
                width="360"
                x="20"
                x1="92"
                x2="92"
                y="350"
                y1="356"
                y2="350"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                type="category"
                width="360"
                x="92"
                y="358"
              >
                <tspan
                  dy="0.71em"
                  x="92"
                >
                  Page B
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                type="category"
                width="360"
                x="20"
                x1="164"
                x2="164"
                y="350"
                y1="356"
                y2="350"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                type="category"
                width="360"
                x="164"
                y="358"
              >
                <tspan
                  dy="0.71em"
                  x="164"
                >
                  Page C
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                type="category"
                width="360"
                x="20"
                x1="236"
                x2="236"
                y="350"
                y1="356"
                y2="350"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                type="category"
                width="360"
                x="236"
                y="358"
              >
                <tspan
                  dy="0.71em"
                  x="236"
                >
                  Page D
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                type="category"
                width="360"
                x="20"
                x1="308"
                x2="308"
                y="350"
                y1="356"
                y2="350"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                type="category"
                width="360"
                x="308"
                y="358"
              >
                <tspan
                  dy="0.71em"
                  x="308"
                >
                  Page E
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                type="category"
                width="360"
                x="20"
                x1="380"
                x2="380"
                y="350"
                y1="356"
                y2="350"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                type="category"
                width="360"
                x="380"
                y="358"
              >
                <tspan
                  dy="0.71em"
                  x="380"
                >
                  Page F
                </tspan>
              </text>
            </g>
          </g>
        </g>
      </svg>
      <div
        class="recharts-tooltip-wrapper"
        style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px; left: 0px;"
      >
        <div
          class="recharts-default-tooltip"
          style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
        >
          <p
            class="recharts-tooltip-label"
            style="margin: 0px;"
          />
        </div>
      </div>
    </div>
    <div
      class="recharts-wrapper"
      style="position: relative; cursor: default; width: 400px; height: 400px;"
    >
      <svg
        class="recharts-surface"
        height="400"
        version="1.1"
        viewBox="0 0 400 400"
        width="400"
      >
        <defs>
          <clippath
            id="recharts4-clip"
          >
            <rect
              height="330"
              width="360"
              x="20"
              y="20"
            />
          </clippath>
        </defs>
        <g
          class="recharts-layer recharts-line"
        >
          <path
            class="recharts-curve recharts-line-curve"
            d="M20,143.75C44,140.3125,68,136.875,92,123.125C116,109.375,140,20,164,20C188,20,212,308.75,236,308.75C260,308.75,284,70.32499999999999,308,70.32499999999999C332,70.32499999999999,356,162.725,380,255.125"
            fill="none"
            height="330"
            points="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
            stroke="#ff7300"
            stroke-dasharray="0px 0px"
            stroke-width="1"
            type="monotone"
            width="360"
          />
        </g>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
        >
          <line
            class="recharts-cartesian-axis-line"
            fill="none"
            height="30"
            orientation="bottom"
            stroke="#666"
            type="category"
            width="360"
            x="20"
            x1="20"
            x2="380"
            y="350"
            y1="350"
            y2="350"
          />
          <g
            class="recharts-cartesian-axis-ticks"
          >
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                type="category"
                width="360"
                x="20"
                x1="20"
                x2="20"
                y="350"
                y1="356"
                y2="350"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                type="category"
                width="360"
                x="20"
                y="358"
              >
                <tspan
                  dy="0.71em"
                  x="20"
                >
                  Page F
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                type="category"
                width="360"
                x="20"
                x1="92"
                x2="92"
                y="350"
                y1="356"
                y2="350"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                type="category"
                width="360"
                x="92"
                y="358"
              >
                <tspan
                  dy="0.71em"
                  x="92"
                >
                  Page E
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                type="category"
                width="360"
                x="20"
                x1="164"
                x2="164"
                y="350"
                y1="356"
                y2="350"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                type="category"
                width="360"
                x="164"
                y="358"
              >
                <tspan
                  dy="0.71em"
                  x="164"
                >
                  Page D
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                type="category"
                width="360"
                x="20"
                x1="236"
                x2="236"
                y="350"
                y1="356"
                y2="350"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                type="category"
                width="360"
                x="236"
                y="358"
              >
                <tspan
                  dy="0.71em"
                  x="236"
                >
                  Page C
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                type="category"
                width="360"
                x="20"
                x1="308"
                x2="308"
                y="350"
                y1="356"
                y2="350"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                type="category"
                width="360"
                x="308"
                y="358"
              >
                <tspan
                  dy="0.71em"
                  x="308"
                >
                  Page B
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                type="category"
                width="360"
                x="20"
                x1="380"
                x2="380"
                y="350"
                y1="356"
                y2="350"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                type="category"
                width="360"
                x="380"
                y="358"
              >
                <tspan
                  dy="0.71em"
                  x="380"
                >
                  Page A
                </tspan>
              </text>
            </g>
          </g>
        </g>
      </svg>
      <div
        class="recharts-tooltip-wrapper"
        style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px; left: 0px;"
      >
        <div
          class="recharts-default-tooltip"
          style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
        >
          <p
            class="recharts-tooltip-label"
            style="margin: 0px;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
